File: src/test/java/uk/gov/justice/digital/hmpps/oauth2server/timed/DeleteDisabledUsersServiceIntTest.java
package uk.gov.justice.digital.hmpps.oauth2server.timed;

import com.microsoft.applicationinsights.TelemetryClient;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase;
import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase.Replace;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.context.annotation.Import;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.junit4.SpringRunner;
import org.springframework.test.context.transaction.TestTransaction;
import org.springframework.transaction.annotation.Transactional;
import uk.gov.justice.digital.hmpps.oauth2server.auth.repository.UserRepository;
import uk.gov.justice.digital.hmpps.oauth2server.auth.repository.UserRetriesRepository;
import uk.gov.justice.digital.hmpps.oauth2server.config.AuthDbConfig;
import uk.gov.justice.digital.hmpps.oauth2server.config.FlywayConfig;
import uk.gov.justice.digital.hmpps.oauth2server.config.NomisDbConfig;

import javax.sql.DataSource;

import static org.assertj.core.api.Assertions.assertThat;

@RunWith(SpringRunner.class)
@DataJpaTest
@ActiveProfiles("dev")
@Import({AuthDbConfig.class, NomisDbConfig.class, FlywayConfig.class})
@AutoConfigureTestDatabase(replace = Replace.NONE)
@Transactional(transactionManager = "authTransactionManager")
public class DeleteDisabledUsersServiceIntTest {
    @Autowired
    private UserRepository userRepository;
    @Autowired
    private UserRetriesRepository userRetriesRepository;
    @Autowired
    @Qualifier("authDataSource")
    private DataSource dataSource;
    @Mock
    private TelemetryClient telemetryClient;

    private DeleteDisabledUsersService service;

    @Before
    public void setUp() {
        service = new DeleteDisabledUsersService(userRepository, userRetriesRepository, telemetryClient);
    }

    @Test
    public void findAndDeleteDisabledAuthUsers_Processed() {
        assertThat(service.processInBatches()).isEqualTo(3);

        TestTransaction.flagForCommit();
        TestTransaction.end();
        TestTransaction.start();

        final var jdbcTemplate = new JdbcTemplate(dataSource);
        final var whereClause = "where username in ('AUTH_DELETE', 'AUTH_DELETEALL', 'NOMIS_DELETE')";

        final var retries = jdbcTemplate.queryForList(String.format("select * from user_retries %s", whereClause));
        assertThat(retries).isEmpty();

        final var tokens = jdbcTemplate.queryForList(String.format("select * from user_token %s", whereClause));
        assertThat(tokens).isEmpty();

        final var users = jdbcTemplate.queryForList(String.format("select * from user_email %s", whereClause));
        assertThat(users).isEmpty();

        final var people = jdbcTemplate.queryForList(String.format("select * from person %s", whereClause));
        assertThat(people).isEmpty();

        final var roles = jdbcTemplate.queryForList(String.format("select * from user_email_roles %s", whereClause));
        assertThat(roles).isEmpty();

        final var groups = jdbcTemplate.queryForList(String.format("select * from user_email_groups %s", whereClause).replace("username", "useremail_username"));
        assertThat(groups).isEmpty();
    }
}

File: src/test/kotlin/uk/gov/justice/digital/hmpps/oauth2server/timed/DeleteDisabledUsersServiceIntTest.kt
package uk.gov.justice.digital.hmpps.oauth2server.timed

import com.microsoft.applicationinsights.TelemetryClient
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.Mock
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest
import org.springframework.context.annotation.Import
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.context.transaction.TestTransaction
import org.springframework.transaction.annotation.Transactional
import uk.gov.justice.digital.hmpps.oauth2server.auth.repository.UserRepository
import uk.gov.justice.digital.hmpps.oauth2server.auth.repository.UserRetriesRepository
import uk.gov.justice.digital.hmpps.oauth2server.config.AuthDbConfig
import uk.gov.justice.digital.hmpps.oauth2server.config.FlywayConfig
import uk.gov.justice.digital.hmpps.oauth2server.config.NomisDbConfig
import javax.sql.DataSource

@Suppress("SqlResolve")
@DataJpaTest
@ActiveProfiles("test")
@Import(AuthDbConfig::class, NomisDbConfig::class, FlywayConfig::class)
@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
@Transactional(transactionManager = "authTransactionManager")
class DeleteDisabledUsersServiceIntTest {
  @Autowired
  private lateinit var userRepository: UserRepository

  @Autowired
  private lateinit var userRetriesRepository: UserRetriesRepository

  @Autowired
  @Qualifier("authDataSource")
  private lateinit var dataSource: DataSource

  @Mock
  private lateinit var telemetryClient: TelemetryClient
  private lateinit var service: DeleteDisabledUsersService

  @BeforeEach
  fun setUp() {
    service = DeleteDisabledUsersService(userRepository, userRetriesRepository, telemetryClient)
  }

  @Test
  fun findAndDeleteDisabledAuthUsers_Processed() {
    val authDeleteId = userRepository.findByUsername("AUTH_DELETE").orElseThrow().id
    val authDeleteAllId = userRepository.findByUsername("AUTH_DELETEALL").orElseThrow().id
    val nomisDeleteId = userRepository.findByUsername("NOMIS_DELETE").orElseThrow().id
    assertThat(service.processInBatches()).isEqualTo(3)
    TestTransaction.flagForCommit()
    TestTransaction.end()
    TestTransaction.start()
    val jdbcTemplate = JdbcTemplate(dataSource)
    val usernameWhereClause = "where username in ('AUTH_DELETE', 'AUTH_DELETEALL', 'NOMIS_DELETE')"
    val userIdWhereClause = "where user_id in ('$authDeleteId', '$authDeleteAllId', '$nomisDeleteId')"
    checkNoRows(jdbcTemplate, "user_retries", usernameWhereClause)
    checkNoRows(jdbcTemplate, "users", usernameWhereClause)
    checkNoRows(jdbcTemplate, "user_token", userIdWhereClause)
    checkNoRows(jdbcTemplate, "user_role", userIdWhereClause)
    checkNoRows(jdbcTemplate, "user_group", userIdWhereClause)
  }

  private fun checkNoRows(jdbcTemplate: JdbcTemplate, table: String, whereClause: String) {
    val retries = jdbcTemplate.queryForList("select * from $table $whereClause")
    assertThat(retries).isEmpty()
  }
}
